protocol/igmp.go

```diff
@@ -0,0 +1,363 @@
+package protocol
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+)
+
+const (
+	IGMPQuery        uint8 = 0x11
+	IGMPv1Report     uint8 = 0x12
+	IGMPv2Report     uint8 = 0x16
+	IGMPv2LeaveGroup uint8 = 0x17
+	IGMPv3Report     uint8 = 0x22
+
+	IGMPIsIn  uint8 = 0x01 // Type MODE_IS_INCLUDE, source addresses x
+	IGMPIsEx  uint8 = 0x02 // Type MODE_IS_EXCLUDE, source addresses x
+	IGMPToIn  uint8 = 0x03 // Type CHANGE_TO_INCLUDE_MODE, source addresses x
+	IGMPToEx  uint8 = 0x04 // Type CHANGE_TO_EXCLUDE_MODE, source addresses x
+	IGMPAllow uint8 = 0x05 // Type ALLOW_NEW_SOURCES, source addresses x
+	IGMPBlock uint8 = 0x06 // Type BLOCK_OLD_SOURCES, source addresses x
+)
+
+type IGMPMessage interface {
+	GetMessageType() uint8
+}
+
+// IGMPv1:
+//    0                   1                   2                   3
+//    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |Version| Type  |    Unused     |           Checksum            |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                         Group Address                         |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//
+// IGMPv2:
+//    0                   1                   2                   3
+//    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |      Type     | Max Resp Time |           Checksum            |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                         Group Address                         |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IGMPv1or2 struct {
+	Type            uint8
+	MaxResponseTime uint8 // It is 0 for IGMPv1 message.
+	Checksum        uint16
+	GroupAddress    net.IP
+}
+
```

Comment:
should we have constructors for these, e.g. `NewIGMPv1(type uint8)`, ` NewIGMPv2(type uint8)`? Or something else that may be easier to consume by client code?

Reply:
added